mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/transformation/modelixToMps/initial/ITreeToSTreeTransformer.kt

```diff
@@ -18,32 +18,56 @@ package org.modelix.mps.sync.transformation.modelixToMps.initial
 
 import mu.KotlinLogging
 import org.modelix.kotlin.utils.UnstableModelixFeature
+import org.modelix.model.api.BuiltinLanguages
 import org.modelix.model.api.IBranch
 import org.modelix.model.api.INode
+import org.modelix.model.api.getRootNode
 import org.modelix.model.mpsadapters.MPSLanguageRepository
 import org.modelix.mps.sync.IBinding
+import org.modelix.mps.sync.tasks.SyncDirection
+import org.modelix.mps.sync.tasks.SyncLock
+import org.modelix.mps.sync.tasks.SyncQueue
 import org.modelix.mps.sync.transformation.modelixToMps.transformers.ModuleTransformer
 import org.modelix.mps.sync.util.isModule
 import org.modelix.mps.sync.util.nodeIdAsLong
 import java.util.Collections
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.atomic.AtomicReference
 
 @UnstableModelixFeature(reason = "The new modelix MPS plugin is under construction", intendedFinalization = "2024.1")
-class ITreeToSTreeTransformer(branch: IBranch, mpsLanguageRepository: MPSLanguageRepository) {
+class ITreeToSTreeTransformer(private val branch: IBranch, mpsLanguageRepository: MPSLanguageRepository) {
 
     private val logger = KotlinLogging.logger {}
 
     private val moduleTransformer = ModuleTransformer(branch, mpsLanguageRepository)
 
-    fun transform(entryPoint: INode): Iterable<IBinding> {
-        require(entryPoint.isModule()) { "Transformation entry point (Node $entryPoint) must be a Module" }
+    private val syncQueue = SyncQueue
 
-        return try {
-            @Suppress("UNCHECKED_CAST")
-            moduleTransformer.transformToModuleCompletely(entryPoint.nodeIdAsLong(), true)
-                .getResult().get() as Iterable<IBinding>
-        } catch (ex: Exception) {
-            logger.error(ex) { "Transformation of Node tree starting from Node $entryPoint failed." }
-            Collections.emptyList()
+    fun transform(moduleId: String): Iterable<IBinding> {
+        val result = syncQueue.enqueue(linkedSetOf(SyncLock.MODELIX_READ), SyncDirection.NONE) {
+            val result = AtomicReference<INode>()
+            branch.runRead {
+                val moduleNode =
+                    branch.getRootNode().allChildren.firstOrNull { it.getPropertyValue(BuiltinLanguages.MPSRepositoryConcepts.Module.id) == moduleId }
+                require(moduleNode != null) { "Module node with ID '$moduleId' is not found on the root level." }
+                result.set(moduleNode)
+            }
+            result.get()
+        }.continueWith(linkedSetOf(SyncLock.MODELIX_READ), SyncDirection.NONE) {
+            val entryPoint = it as INode
+            require(entryPoint.isModule()) { "Transformation entry point (Node $entryPoint) must be a Module" }
```

Comment:
nit: This can be checked earlier

Reply:
fixed in c7fbe666f23b477e94413e7edabeb1f66be3ff28